File: baekjoon/7576.cpp
#include <iostream>
#include <vector>
#include <queue>
using namespace std;
int M, N;
int** box;
int** visited;

// 토마토

struct tomato {
	int x;
	int y;
};

int main()
{
	cin >> M >> N;
	queue <tomato> q;
	box = new int* [1001];
	visited = new int* [1001];

	for (int i = 0; i < 1001; i++)
	{
		box[i] = new int[1001];
		visited[i] = new int[1001];
	}

	for (int i = 0; i < N; i++)
		for (int j = 0; j < M; j++)
		{
			visited[i][j] = -1;
			cin >> box[i][j];
			if (box[i][j] == 1) // 만약 상자에 토마토가 있다면
			{
				q.push({ i, j }); // 주변 토마토가 익을 수 있으므로 push
				visited[i][j] = 0; // 토마토가 있는 영역 방문 표시
			}
		}


	tomato dir[4] = { {1,0}, {-1,0}, {0,1}, {0,-1} };

	while (!q.empty())
	{
		tomato now = q.front();
		q.pop();

		for (int i = 0; i < 4; i++)
		{
			int tx = now.x + dir[i].x;
			int ty = now.y + dir[i].y;

			if (tx < 0 || tx >= N)
				continue;

			if (ty < 0 || ty >= M)
				continue;

			if (box[tx][ty] == 0 && visited[tx][ty] == -1) // 토마토가 익지 않았고 방문하지 않았을 때
			{
				visited[tx][ty] = visited[now.x][now.y] + 1; // 이동한 횟수 더하기
				q.push({ tx,ty });
			}
		}
	}

	int answer = 0;
	for (int i = 0; i < N; i++)
		for (int j = 0; j < M; j++)
		{
			if (answer < visited[i][j])
				answer = visited[i][j];
		}

	for (int i = 0; i < N; i++)
		for (int j = 0; j < M; j++)
		{
			if (box[i][j] == 0)
				if (visited[i][j] == -1) // 만약 익지 않은 토마토가 하나라도 존재한다면
				{
					cout << -1;
					return 0;
				}
		}

	cout << answer;

	return 0;
}

File: baekjoon/7576.kt
import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.*

// 토마토

var N: Int = 0
var M: Int = 0
lateinit var tomato: Array<IntArray>

fun isEnd(): Boolean {
    for (i in 0 until N)
        for (j in 0 until M)
            if (tomato[i][j] == 0)
                return false

    return true
}

fun main() {
    val br = BufferedReader(InputStreamReader(System.`in`))
    var token = StringTokenizer(br.readLine())
    M = token.nextToken().toInt()
    N = token.nextToken().toInt()
    tomato = Array(N) { IntArray(M) }
    val que: Queue<Pair<Int, Int>> = LinkedList()

    for (i in 0 until N) {
        token = StringTokenizer(br.readLine())
        for (j in 0 until M) {
            tomato[i][j] = token.nextToken().toInt()
            if (tomato[i][j] == 1)
                que.offer(Pair(i, j))
        }
    }

    val dir: Array<Pair<Int, Int>> = arrayOf(Pair(0, 1), Pair(1, 0), Pair(0, -1), Pair(-1, 0))
    var isPossible = false
    var answer = 0

    if (!isEnd()) {
        while (!que.isEmpty()) {
            val front = que.poll()
            for (d in dir) {
                val r = front.first + d.first
                val c = front.second + d.second

                if (r < 0 || r >= N || c < 0 || c >= M) continue
                if (tomato[r][c] == -1) continue
                if (tomato[r][c] == 0) {
                    tomato[r][c] = tomato[front.first][front.second] + 1

                    if (answer < tomato[r][c])
                        answer = tomato[r][c]

                    que.offer(Pair(r, c))
                }
            }
        }

        if (isEnd()) {
            answer--
            isPossible = true
        }
    } else {
        answer = 0
        isPossible = true
    }

    if (!isPossible)
        answer = -1

    print(answer)
}